### DISCUSSION TITLE
How to merge the outputs of two queries?

### DISCUSSION THREAD

**Discussion by egor83** [sqlalchemy/sqlalchemy#6883] [Usage Questions] [answered]
- SQLAlchemy version 1.4.22;
- Flask version 1.1.2;
- Flask-SQLAlchemy version 2.5.1;
- PostgreSQL version 12.7;

I'm working with the following table:
```python
from sqlalchemy.dialects.postgresql import ARRAY
 
 
class CaseStudy(db.Model):
    __tablename__ = "case_studies"
    id = db.Column(db.Integer, primary_key=True)
    ...
    solutions = db.Column(ARRAY(db.Text))
    challenge_who = db.Column(db.Text)
    ...
```

and I have two queries:
```python
# where q is the search argument
query1 = session.query(cls.id).select_from(func.unnest(cls.solutions).alias("sols")).filter(column("sols").ilike(f"%{q}%"))
query2 = session.query(cls.id).filter(cls.challenge_who.ilike(f"%{q}%"))
```

Now I want to merge the outputs of these two queries (for ordering, pagination etc), but so far I haven't been able to.

What I tried:

```python
subquery = query2.subquery()
 
result = query1.join(subquery)
# sqlalchemy.exc.InvalidRequestError: Don't know how to join to
# <sqlalchemy.sql.selectable.Subquery at 0x7f0d2adb0890; anon_1>. Please use the .select_from() method to 
# establish an explicit left side, as well as providing an explicit ON clause if not present already to help 
# resolve the ambiguity.
 
result = query1.subquery().join(query2.subquery(), query1.id == query2.id)
# AttributeError: 'BaseQuery' object has no attribute 'id'
 
q1 = aliased(query1)
q2 = aliased(query2)
result = query1.join(cls, cls.id==q1.id).join(query2.subquery())
# AttributeError: 'BaseQuery' object has no attribute 'mapper'
 
result = join(query1, subquery)
# sqlalchemy.exc.NoForeignKeysError: Can't find any foreign key relationships between 'anon_1' and 'anon_1'.

result = join(query1, query2)
# sqlalchemy.exc.NoForeignKeysError: Can't find any foreign key relationships between 'anon_1' and 'anon_1'.
```

Sorry I didn't save full stack traces for errors, can add these if needed.

Any advice on how to do this?

BTW, if anyone has a better solution for query1, that'd be much appreciated too! (more details [here](https://stackoverflow.com/q/68742752/118520)).

**Comment by zzzeek** [comment, marked as answer]
so you are using the ORM Query object, which is [legacy](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-unified-with-core-select), in all kinds of ways for which you normally use select() and subquery() objects.  so if you use [2.0 style](https://docs.sqlalchemy.org/en/14/glossary.html#term-1) querying, you will at least get better error messages when you do the wrong thing (such as `query1.id`).

Also, in SQL, the usual way to "merge" queries is to use UNION ALL.

Some background on the API usage:

* aliased() is an ORM function that receives an ORM mapped class, like `CaseStudy`, as the first argument.  you would not send a query or select() object to this as the first argument.    
* the line `result = query1.subquery().join(query2.subquery(), query1.id == query2.id)` is wrong because you are not using the "subquery" object you're trying to create, and the query1 is against "cls.id" so that is the other side of the ON clause, so it should be something like:

```py
   sq = query2.subquery()
   result = query1.join(sq, cls.id == sq.c.id)
```

I think the first thing you need here is to know what SQL you want to emit and it seems like you would want a UNION, if you are trying to have rows from query1 first, then rows from query2.

**Reply by egor83** [reply]
Thanks for reply, @zzzeek!

Re: migration to 2.0, it seems we'd need to rewrite all queries in our project, right? That's quite a few, so if possible, we'd rather do it later.

Re: this snippet `result = query1.join(sq, cls.id == sq.c.id)`, here's what I got: 
`HINT:  There is an entry for table "case_studies", but it cannot be referenced from this part of the query.`

Simple UNION of two queries worked, so this was the SQL:
```sql
SELECT case_studies.id AS case_studies_id 
FROM case_studies, unnest(case_studies.solutions) AS sols 
WHERE sols ILIKE 'd'
UNION
SELECT case_studies.id AS case_studies_id 
FROM case_studies 
WHERE case_studies.challenge_who ILIKE 'who';
```

And this was the SQLAlchemy:
```python
result = query1.union(query2)
```

First quick tests show that it works as needed. Thanks a lot for you help!

**Reply by egor83** [reply]
Another problem - when I tried to order the merged result, I got the following error:
`sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedTable) missing FROM-clause entry for table "case_studies"`

```sql
[SQL: SELECT anon_1.case_studies_id AS anon_1_case_studies_id 
FROM (SELECT case_studies.id AS case_studies_id 
FROM case_studies, unnest(case_studies.solutions) AS sols 
WHERE sols ILIKE %(sols_1)s UNION SELECT case_studies.id AS case_studies_id 
FROM case_studies 
WHERE case_studies.challenge_who ILIKE %(challenge_who_1)s OR case_studies.challenge_what ILIKE %(challenge_what_1)s OR case_studies.challenge_why ILIKE %(challenge_why_1)s) AS anon_1 ORDER BY case_studies.created_at]
[parameters: {'sols_1': '%d%', 'challenge_who_1': '%d%', 'challenge_what_1': '%d%', 'challenge_why_1': '%d%'}]
```

I see it wraps first SELECT in another SELECT, and this seems to cause the problem - any idea how to handle this?

**Reply by egor83** [reply]
PS Re: migration to 2.0, we're using Flask-SQLAlchemy, and quick googling doesn't show how to enable future for it.

**Reply by egor83** [reply]
Solved by adding `created_at` to initial queries.

**Reply by egor83** [reply]
Thanks again!

**Reply by zzzeek** [reply]
great, UNIONs are a little awkward in particular with the old interface.      you can use 2.0 style querying right now for any arbitrary query, there's no need to migrate all your code or to use the future flag, `session.execute()` in 1.4 can run `select()` objects directly and return ORM results.

**Reply by egor83** [reply]
Yeah, UNION worked out of the box.

And thanks for the pointer! Will try migrating them one-by-one then.

**Reply by egor83** [reply]
Re: new-style queries: this query worked fine (returned model objects, as expected):
```python
result_stmt = select(cls)
result = session.execute(result_stmt).scalars().all()
```

But the first one now returns just a list of IDs (instead of list of model objects which it returned with the old-style query):
```python
query1 = (
    session.query(cls)
    .select_from(func.unnest(cls.solutions).alias("sols"))
    .filter(column("sols").ilike(f"%{q}%"))
)
query2 = session.query(cls).filter(
    cls.challenge_who.ilike(f"%{q}%")
    | cls.challenge_what.ilike(f"%{q}%")
    | cls.challenge_why.ilike(f"%{q}%")
)
query_all_stmt = query1.union(query2)
result = session.execute(query_all_stmt).scalars().all()
```

SQL for this query is as follows:
```sql
SELECT anon_1.case_studies_id AS anon_1_case_studies_id, [...]
FROM (SELECT case_studies.id AS case_studies_id, [...]
FROM case_studies, unnest(case_studies.solutions) AS sols 
WHERE sols ILIKE %(sols_1)s UNION SELECT [...])
```
so we are selecting not from `case_studies`, but from `anon_1` - could that be the reason? And in any case, any advice how to get model objects from this query?

**Reply by zzzeek** [reply]
yes in order to get objects from a union you have to use the aliased construct, new tutorial introduces this at https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#orm-entity-aliases, however I think the union case should have a section in the querying guide at https://docs.sqlalchemy.org/en/14/orm/queryguide.html#select-statements, it looks like this:

```py
s1 = select(A).where(A.id == 1)
s2 = select(A).where(A.id == 2)

a_union = aliased(A, s1.union(s2).subquery())
get_union_as = select(a_union)

a_recs = s.execute(get_union_as).scalars().all()
print(a_recs)
```

this basically means in "new style" we manually lay out what the old Query did automatically.   this definitely needs more docs because even I've been tripping over this recently.

**Reply by egor83** [reply]
Thanks again, you've been a great help!
